aten/src/ATen/cudnn/Handle.cpp

```diff
@@ -34,8 +35,67 @@ struct Handle {
 };
 
 std::mutex mutex;
-std::unordered_map<int, Handle> handles;
 
+// Handles are lazily created as different threads request them,
+// but are never destroyed until the end of the program.
+// This is the desired behavior: We want unused handles to stick around,
+// so they can be reused by other threads later.
+//
+// To prevent potential deadlocks, we explicitly choose not to cap the number
+// of handles that are created per device.
+// Example of danger: If we cap the max handles at 4, and 5 threads are sharing a device,
+// only 4 can make forward progress at any time. The other 4 will not release their
+// handles until they exit, so the fifth cannot make progress until then.  This is
+// not a problem...UNLESS all 5 threads attempt some sort of synchronization at an
+// intermediate point (ie, before any of them have exited).  We have no way to anticipate
+// or enforce that user threads will not attempt such intermediate synchronization.
+// The only way to ensure safety is to avoid imposing a cap on the number of handles.
+std::unordered_map<int, std::stack<Handle>> created_handles;
+std::unordered_map<int, std::stack<cudnnHandle_t>> available_handles;
+
+struct PoolWindow
+{
+  std::unordered_map<int, cudnnHandle_t> my_handles;
+
+  PoolWindow(){}
+
+  cudnnHandle_t reserve(int device)
+  {
+    // If this thread already has a handle for this device, return it
+    if(my_handles.find(device) != my_handles.end())
+      return my_handles[device];
+
+    // otherwise, either grab a handle from the pool if one is available,
+    // or if not, create a new one.
+    std::lock_guard<std::mutex> guard(mutex);
+
+    if(available_handles[device].size() > 0)
+    {
+      my_handles[device] = available_handles[device].top();
+      available_handles[device].pop();
+    }
+    else
+    {
+      created_handles[device].emplace(); // no arguments to Handle constructor
+      my_handles[device] = created_handles[device].top().handle;
+    }
+
+    return my_handles[device];
+  }
+
+  void release()
```

Comment:
Make this method private? Otherwise, you need to correctly clear `available_handles` when you're done or you'll double-release handles.

Reply:
Aye, this is only ever intended to be called by the destructor.